bottlecap/src/bin/bottlecap/main.rs

```diff
@@ -329,33 +330,44 @@ async fn main() -> Result<()> {
         .await
         .map_err(|e| Error::new(std::io::ErrorKind::InvalidData, e.to_string()))?;
 
-    if let Some(resolved_api_key) =
-        resolve_secrets(Arc::clone(&config), &aws_config, &mut aws_credentials).await
+    let aws_config = Arc::new(aws_config);
+    let aws_credentials = Arc::new(RwLock::new(aws_credentials));
+    let api_key_factory = {
+        let config = Arc::clone(&config);
+        let aws_config = Arc::clone(&aws_config);
+        let aws_credentials = Arc::clone(&aws_credentials);
+
+        Arc::new(ApiKeyFactory::new_from_resolver(Arc::new(move || {
+            let config = Arc::clone(&config);
+            let aws_config = Arc::clone(&aws_config);
+            let aws_credentials = Arc::clone(&aws_credentials);
+
+            Box::pin(async move {
+                resolve_secrets(config, aws_config, aws_credentials)
+                    .await
+                    .expect("Failed to resolve API key")
```

Comment:
Any better way to handle this?

Reply:
It will panic at flush time and stop the runtime, so I wonder if there's a way for the extension to stop gracefully at that time without stopping the runtime.